vowpalwabbit/global_data.cc

```diff
@@ -167,7 +167,20 @@ void vw::learn(example& ec)
   if (ec.test_only || !training)
     VW::LEARNER::as_singleline(l)->predict(ec);
   else
-    VW::LEARNER::as_singleline(l)->learn(ec);
+  {
+    if (l->predict_before_learn)
+    {
+      VW::LEARNER::as_singleline(l)->predict(ec);
+      copy_prediction(ec.pred);
+      auto restore_guard = VW::scope_exit([&ec, this]
+      {
+        std::swap(ec.pred, _predict_buffer);
```

Comment:
If `_predict_buffer` is only used in the predict and learn functions it really should be a local variable.

Reply:
Having a non-local variable prevents allocations in steady state.

Reply:
It's really quite a small allocation for this and in an effort to avoid it you are adding another field (union at that) to the god object we are trying to get rid of/reduce

Reply:
It's possible to have predictions with many elements.  With cb_adf_explore, ccb, and continuous actions it's possible to have non-trivial amount of elements in predictions.  This pattern is used in many reductions to avoid allocations. 

_predict_buffer is used by vw for coordinating between learn, predict and finish_example at a driver level.  This is different than adding something to struct vw that is use by one or even a couple of reductions.  It seems appropriate to add it here.

Reply:
Without a performance test showing that this is required and the effects of this in real world use I still feel opposed to this driver level state.

Reply:
I understand the nature of objection.  I will setup a session Monday to review the contentious changes.   I will add this to the discussion list.  Does that work?